streams/src/test/java/org/apache/kafka/streams/integration/HeadersSupportIntegrationTest.java

```diff
@@ -0,0 +1,239 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.integration;
+
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.header.Header;
+import org.apache.kafka.common.header.Headers;
+import org.apache.kafka.common.header.internals.RecordHeader;
+import org.apache.kafka.common.header.internals.RecordHeaders;
+import org.apache.kafka.common.serialization.LongDeserializer;
+import org.apache.kafka.common.serialization.LongSerializer;
+import org.apache.kafka.common.serialization.Serdes;
+import org.apache.kafka.streams.KafkaStreams;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.StreamsBuilder;
+import org.apache.kafka.streams.StreamsConfig;
+import org.apache.kafka.streams.Topology;
+import org.apache.kafka.streams.integration.utils.EmbeddedKafkaCluster;
+import org.apache.kafka.streams.integration.utils.IntegrationTestUtils;
+import org.apache.kafka.streams.processor.AbstractProcessor;
+import org.apache.kafka.streams.processor.Processor;
+import org.apache.kafka.streams.processor.ProcessorSupplier;
+import org.apache.kafka.test.IntegrationTest;
+import org.apache.kafka.test.StreamsTestUtils;
+import org.apache.kafka.test.TestUtils;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Properties;
+import java.util.concurrent.ExecutionException;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+@Category({IntegrationTest.class})
+public class HeadersSupportIntegrationTest {
+    private static final int NUM_BROKERS = 3;
+
+    @ClassRule
+    public static final EmbeddedKafkaCluster CLUSTER = new EmbeddedKafkaCluster(NUM_BROKERS, new Properties() {
+        {
+            put("auto.create.topics.enable", false);
+        }
+    });
+
+    private static String applicationId;
+    private final static String CONSUMER_GROUP_ID = "headersConsumer";
+    private static final String INPUT_TOPIC = "inputTopic";
+    private static final String THROUGH_TOPIC = "throughTopic";
+    private static final String OUTPUT_TOPIC = "outputTopic";
+
+
+    private int testNumber = 0;
+
+    @Before
+    public void createTopics() throws InterruptedException {
+        applicationId = "appId-" + ++testNumber;
+        CLUSTER.deleteTopicsAndWait(
+                INPUT_TOPIC,
+                THROUGH_TOPIC,
+                OUTPUT_TOPIC);
+
+        CLUSTER.createTopics(INPUT_TOPIC, THROUGH_TOPIC, OUTPUT_TOPIC);
+    }
+
+    @Test
+    public void shouldCopyHeadersToIntermediateAndSinkTopic() throws ExecutionException, InterruptedException {
```

Comment:
nit: simplify to `throw Exception`
for tests, it's irrelevant what specific exceptions are throws (same below)